x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1141,19 +1161,37 @@ private static boolean supportsStringImplicitCasting(DataType type) {
             return type == DATETIME || type == IP || type == VERSION || type == BOOLEAN;
         }
 
-        public static Expression castStringLiteral(Expression from, DataType target) {
+        private static UnresolvedAttribute unresolvedAttribute(Expression value, String type, Exception e) {
+            String message = format(
+                "Cannot convert string [{}] to [{}], error [{}]",
+                value.fold(),
+                type,
+                (e instanceof ParsingException pe) ? pe.getErrorMessage() : e.getMessage()
+            );
+            return new UnresolvedAttribute(value.source(), String.valueOf(value.fold()), message);
+        }
+
+        private static Expression castStringLiteralToTemporalAmount(Expression from) {
+            try {
+                TemporalAmount result = maybeParseTemporalAmount(from.fold().toString().strip());
+                if (result == null) {
```

Comment:
Is `null` acceptable in any case here?
I find reading the flow of calls a bit "uneven": `castString...()` will try to `maybeParse...()`, which if it returns `null` is OK, but if it can't, it'll throw. Can `maybeParse...()` throw on empty value? Do we need to "maybe" parse?

Reply:
`maybeParseTemporalAmount` might return a null, if the literal string does not follow the format of a temporal(there is space between integer and unit), for example "1", "1w", "" etc. It doesn't throw on empty value. The verification of a non temporal format string is deferred to `resolveType()` to return a proper message. Here are a few examples that the error message is from `resolveType()`:

`stats max(emp_no) by bucket(emp_no, "5")` returns `second argument of [bucket(emp_no, "5")] must be [numeric], found value ["5"] type [keyword]`

`stats max(emp_no) by bucket(hire_date, "5")` returns `second argument of [bucket(hire_date, "5")] must be [integral, date_period or time_duration], found value ["5"] type [keyword]`

`stats max(emp_no) by bucket(hire_date, "5w")` returns `second argument of [bucket(hire_date, "5w")] must be [integral, date_period or time_duration], found value ["5w"] type [keyword]`

`stats max(emp_no) by bucket(hire_date, "")` returns `second argument of [bucket(hire_date, "")] must be [integral, date_period or time_duration], found value [""] type [keyword]`

Reply:
> It doesn't throw on empty value.

Yes, I was wondering if it could. 

The (small, can be ignored ;) ) issue I see is that `maybeParseTemporalAmount()` is only invoked if `isTemporalAmount()`. Meaning that _it has to be_ one. 
But the failure case is not treated uniformly:
* If it's not well formatted because there's just one token (or zero, but that's just theoretical, I think, b/c of the grammar), it returns a `null`. 
* If it's not well formatted because the first token isn't a number, or the second one isn't a defined unit, it'll throw.

And I'm not sure why, since also in the former case, as written, it has to be a temporal amount and it cannot be just one token (or empty).